### PR TITLE
Added literal kind properties for each node.

### PR DESCRIPTION
**Pull request by rbuckton**
This PR adds literal `kind` properties on each `Node` subtype in the compiler. While this is not a full migration to ADTs, it has caught a few errors in the compiler not presently caught by existing unit tests.

This also relaxes the `typeOperatorSpacingRule` for tslint so that it allows union and intersection types to span multiple lines.

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -1111,7 +1111,7 @@ namespace ts {
             }
             else {
                 forEachChild(node, bind);
-                if (node.operator === SyntaxKind.PlusEqualsToken || node.operator === SyntaxKind.MinusMinusToken) {
+                if (node.operator === SyntaxKind.PlusPlusToken || node.operator === SyntaxKind.MinusMinusToken) {
```
Comment (rbuckton) [inline src/compiler/binder.ts]: This was discovered due to the use of the more specific `kind` property.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3886,7 +3886,7 @@ namespace ts {
             if (!links.declaredType) {
                 const enumType = <EnumType>getDeclaredTypeOfEnum(getParentOfSymbol(symbol));
                 links.declaredType = enumType.flags & TypeFlags.Union ?
-                    enumType.memberTypes[getEnumMemberValue(<EnumDeclaration>symbol.valueDeclaration)] :
+                    enumType.memberTypes[getEnumMemberValue(<EnumMember>symbol.valueDeclaration)] :
```
Comment (rbuckton) [inline src/compiler/checker.ts]: This was discovered due to the use of the more specific `kind` property.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -19966,7 +19968,7 @@ namespace ts {
                     currentKind = SetAccessor;
                 }
                 else {
-                    Debug.fail("Unexpected syntax kind:" + prop.kind);
+                    Debug.fail("Unexpected syntax kind:" + (<Node>prop).kind);
```
Comment (rbuckton) [inline src/compiler/checker.ts]: This branch is possibly unnecessary since `prop` is typed as `never` here.

**Comment by rbuckton** [comment]
// CC: @mhegazy, @ahejlsberg, @DanielRosenwasser

File: src/compiler/core.ts [hunk 4]
```diff
@@ -1813,9 +1813,9 @@ namespace ts {
 
     export interface ObjectAllocator {
         getNodeConstructor(): new (kind: SyntaxKind, pos?: number, end?: number) => Node;
-        getTokenConstructor(): new (kind: SyntaxKind, pos?: number, end?: number) => Token;
-        getIdentifierConstructor(): new (kind: SyntaxKind, pos?: number, end?: number) => Token;
-        getSourceFileConstructor(): new (kind: SyntaxKind, pos?: number, end?: number) => SourceFile;
+        getTokenConstructor(): new <TKind extends SyntaxKind>(kind: TKind, pos?: number, end?: number) => Token<TKind>;
+        getIdentifierConstructor(): new (kind: SyntaxKind.Identifier, pos?: number, end?: number) => Identifier;
```
Comment (DanielRosenwasser) [inline src/compiler/core.ts]: Kind of funny that the `Identifier` constructor needs to take a `SyntaxKind`

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -17419,9 +17419,12 @@ namespace ts {
                     }
                 }
 
-                checkCollisionWithCapturedThisVariable(node, node.name);
-                checkCollisionWithRequireExportsInGeneratedCode(node, node.name);
-                checkCollisionWithGlobalPromiseInGeneratedCode(node, node.name);
+                if (isIdentifier(node.name)) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Was this a bug, or are we just trying to assert here that this is an `Identifier`?

File: src/compiler/core.ts (hunk 4 above)
Comment (rbuckton) [inline src/compiler/core.ts]: Who knows, we could eventually have different kind's of `Identifier` nodes.

**Pull request merged**